server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeRole.java

```diff
@@ -132,6 +132,27 @@ public String toString() {
             return Setting.boolSetting("node. " + roleName(), false, Setting.Property.NodeScope);
         }
 
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            UnknownRole that = (UnknownRole) o;
+            return Objects.equals(roleName(), that.roleName()) &&
+                Objects.equals(roleNameAbbreviation(), that.roleNameAbbreviation());
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(roleName(), roleNameAbbreviation());
+        }
+
+        @Override
+        public String toString() {
+            return "UnknownRole{" +
+                "roleName='" + roleName() + '\'' +
+                ", roleAbbreviation='" + roleNameAbbreviation() + '\'' +
+                '}';
+        }
```

Comment:
I think these can all go in the base class, as final, with adding getClass() to the toString implementation?